app/src/main/java/io/apicurio/registry/config/RegistryConfigServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.config;
+
+import static io.quarkus.scheduler.Scheduled.ConcurrentExecution.SKIP;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.slf4j.Logger;
+
+import io.apicurio.registry.mt.TenantContext;
+import io.apicurio.registry.storage.RegistryStorage;
+import io.apicurio.registry.types.Current;
+import io.quarkus.scheduler.Scheduled;
+
+/**
+ * @author eric.wittmann@gmail.com
+ */
+@ApplicationScoped
+public class RegistryConfigServiceImpl implements RegistryConfigService {
+
+    private static final String DEFAULT_TENANT_ID = "___";
+
+    @Inject
+    Logger log;
+
+    @Inject
+    @Current
+    RegistryStorage storage;
+
+    @Inject
+    TenantContext tenantContext;
+
+    private Map<RegistryConfigProperty, Object> globalPropertyCache = new HashMap<>();
+    private Map<String, Map<RegistryConfigProperty, Object>> tenantPropertyCaches = new HashMap<>();
+    private Instant lastRefresh = null;
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#get(io.apicurio.registry.config.RegistryConfigProperty)
+     */
+    @Override
+    public String get(RegistryConfigProperty property) {
+        return get(property, String.class);
+    }
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#getOptional(io.apicurio.registry.config.RegistryConfigProperty)
+     */
+    @Override
+    public Optional<String> getOptional(RegistryConfigProperty property) {
+        return getOptional(property, String.class);
+    }
+
+    /**
+     * @see io.apicurio.registry.config.RegistryConfigService#get(io.apicurio.registry.config.RegistryConfigProperty, java.lang.Class)
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T get(RegistryConfigProperty property, Class<T> propertyType) {
+        // Possibly check the tenant property cache for a value.  If found return it.
+        T rval = property.isEditable() ? (T) tenantPropertyCache().get(property) : null;
+
+        // If not found, check for a global property.
+        if (rval == null) {
+            rval = (T) globalPropertyCache.computeIfAbsent(property, (key) -> {
+                Optional<? extends Object> value = ConfigProvider.getConfig().getOptionalValue(property.propertyName(), propertyType);
+                if (value.isPresent()) {
+                    return (T) value.get();
+                } else {
+                    return (T) property.defaultValue();
```

Comment:
I think if defaultValue is null the logic inside `computeIfAbsent` will be always executed, because never is going to be a value

Reply:
Ah yes - I didn't know this.  Just read the javadoc and you are quite right.  I will fix.